services/src/main/java/org/keycloak/authentication/authenticators/broker/IdpUsernamePasswordForm.java

```diff
@@ -22,51 +22,76 @@
+ * In this case username field will be pre-filled (but still changeable).
  *
  * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
  */
 public class IdpUsernamePasswordForm extends UsernamePasswordForm {
 
     @Override
     protected Response challenge(AuthenticationFlowContext context, MultivaluedMap<String, String> formData) {
-        UserModel existingUser = AbstractIdpAuthenticator.getExistingUser(context.getSession(), context.getRealm(), context.getAuthenticationSession());
-
-        return setupForm(context, formData, existingUser)
+        return setupForm(context, formData, getExistingUser(context))
                 .setStatus(Response.Status.OK)
                 .createLoginUsernamePassword();
     }
 
     @Override
     protected boolean validateForm(AuthenticationFlowContext context, MultivaluedMap<String, String> formData) {
-        UserModel existingUser = AbstractIdpAuthenticator.getExistingUser(context.getSession(), context.getRealm(), context.getAuthenticationSession());
-        context.setUser(existingUser);
+        Optional<UserModel> existingUser = getExistingUser(context);
+        existingUser.ifPresent(context::setUser);
+
+        boolean result = validateUserAndPassword(context, formData);
 
         // Restore formData for the case of error
         setupForm(context, formData, existingUser);
 
-        return validatePassword(context, existingUser, formData);
+        return result;
     }
 
-    protected LoginFormsProvider setupForm(AuthenticationFlowContext context, MultivaluedMap<String, String> formData, UserModel existingUser) {
+    protected LoginFormsProvider setupForm(AuthenticationFlowContext context, MultivaluedMap<String, String> formData, Optional<UserModel> existingUser) {
         SerializedBrokeredIdentityContext serializedCtx = SerializedBrokeredIdentityContext.readFromAuthenticationSession(context.getAuthenticationSession(), AbstractIdpAuthenticator.BROKERED_CONTEXT_NOTE);
         if (serializedCtx == null) {
             throw new AuthenticationFlowException("Not found serialized context in clientSession", AuthenticationFlowError.IDENTITY_PROVIDER_ERROR);
         }
 
-        formData.putSingle(AuthenticationManager.FORM_USERNAME, existingUser.getUsername());
-        return context.form()
+        existingUser.ifPresent(u -> formData.putSingle(AuthenticationManager.FORM_USERNAME, u.getUsername()));
+
+        LoginFormsProvider form = context.form()
                 .setFormData(formData)
-                .setAttribute(LoginFormsProvider.USERNAME_EDIT_DISABLED, true)
-                .setInfo(Messages.FEDERATED_IDENTITY_CONFIRM_REAUTHENTICATE_MESSAGE, existingUser.getUsername(), serializedCtx.getIdentityProviderId());
+                .setAttribute(LoginFormsProvider.REGISTRATION_DISABLED, true)
+                .setInfo(Messages.FEDERATED_IDENTITY_CONFIRM_REAUTHENTICATE_MESSAGE, serializedCtx.getIdentityProviderId());
```

Comment:
There is one small issue that changing the message and arguments would mean that message will be a bit broken for other languages despite english. I see you updated english resource bundle in this PR, but assuming you did not updated it, the message will be something like `Authenticate as 'twitter' to link your account with`
Some similar out-of-sense message is now likely displayed for other languages.

I wonder if you can try to update the message for other languages as well? I understand you don't understand other languages, but maybe at least some best-effort? :)

For example, I assume that French message `Identifiez vous en tant que {0} afin de lier votre compte avec {1}`
could be probably changed to something like `Identifiez vous en tant afin de lier votre compte avec {0}`
(I don't understand french, so really just best-effort from my side :)

Another option is to remove the message and create completely new message, which will mean that other languages will see just default english message.

Reply:
I have managed to provide translations for all languages, except `ja`, `zh_CN`, `tr` and `no`. For these languages, I have commented out the `federatedIdentityConfirmReauthenticateMessage` key until we have proper translations.